### PR TITLE
Bug 1082052 - [Messages][Refactoring] Return a Promise from SMIL.pars…

### PR DESCRIPTION
**Pull request by julienw**
…e r=azasypkin

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2e536804661e27fcf7dae4acc4347614ed5afaee

File: apps/sms/views/shared/js/smil.js [hunk 1]
```diff
@@ -365,9 +369,10 @@ window.SMIL = {
     if (!smil || attachmentsNotFound || !slides.length || smilMismatched) {
       // reset slides in the attachments not found case
       slides = Array(attachments.length);
-      attachments.forEach(SMIL_parseWithoutSMIL);
+      result = Promise.all(attachments.map(SMIL_parseWithoutSMIL));
     }
-    exitPoint();
+
+    return (result || Promise.resolve()).then(() => slides);
```
Comment (julienw) [inline apps/sms/views/shared/js/smil.js]: I know I could have done better, especially a better use of Promise.reject to handle errors and fallbacks. As a first step I wanted to get the API right and update unit tests, so that I can use a Promise API in the new code. I can file a separate bug to make this better if you think it's worth it?

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=0a547c52b2dd03864d386a744d7bb835818868d8

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=60e37d5c0f50c49a85a1da30e9e6956be3044b4e

File: apps/sms/views/shared/test/unit/activity_handler_test.js [hunk 2]
```diff
@@ -525,6 +528,59 @@ suite('ActivityHandler', function() {
...
+      SMIL.mParsed = [{ text: 'some text' }];
+
+      simulateMessageReceived(message).then(() => {
+        sinon.assert.calledWithMatch(
+          NotificationHelper.send,
+          sinon.match.string, { body: 'some text' }
+        );
+      }).then(done, done);
+    });
+  });
```
Comment (julienw) [inline apps/sms/views/shared/test/unit/activity_handler_test.js]: looks like we had no test for received MMS :/

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=24ac776bbbe8a0e1ad2fcf66043c4d3cd6914fd8

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a499339055194f9bb7bca7929e0f33d5da83b492

File: apps/sms/desktop-mock/navigator_moz_mobilemessage.js [hunk 3]
```diff
@@ -1304,32 +1304,33 @@
   //  - id: Number specifying the message to retrieve
   //  Returns: request object
   MockNavigatormozMobileMessage.getMessage = function(id) {
-    var request = {
```
Comment (azasypkin) [inline apps/sms/desktop-mock/navigator_moz_mobilemessage.js]: Is this change related to SMIL somehow? :)

File: apps/sms/views/conversation/js/conversation.js [hunk 4]
```diff
@@ -1763,22 +1763,22 @@ var ConversationView = {
...
-      });
+      result = SMIL.parse(message).then(
+        (slideArray) => this.createMmsContent(slideArray)
+      ).then(
+        (mmsContent) => pElement.appendChild(mmsContent)
+      );
     }
 
-    return Promise.resolve(messageDOM);
+    return (result || Promise.resolve()).then(() => messageDOM);
```
Comment (azasypkin) [inline apps/sms/views/conversation/js/conversation.js]: IMO previous two separate ```return``` statements look more readable than ```result || Promise.resolve()).then(() => messageDOM);```, but up to you :)

File: apps/sms/views/conversation/js/conversation.js [hunk 5]
```diff
@@ -1763,22 +1763,22 @@ var ConversationView = {
...
   },
 
   mmsContentParser: function conv_mmsContentParser(message) {
```
Comment (azasypkin) [inline apps/sms/views/conversation/js/conversation.js]: Is this method used anywhere after your change?

File: apps/sms/views/shared/js/smil.js [hunk 6]
```diff
@@ -1,6 +1,3 @@
-/* -*- Mode: js; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- /
```
Comment (azasypkin) [inline apps/sms/views/shared/js/smil.js]: This file deserves one more separate refactoring bug :)

File: apps/sms/views/shared/js/smil.js [hunk 7]
```diff
@@ -238,23 +229,25 @@ window.SMIL = {
...
+          defer.resolve();
         });
       };
       reader.onerror = function(event) {
-        activeReaders--;
         console.error('Error reading text blob');
-        exitPoint();
+        defer.resolve();
       };
       activeReaders++;
```
Comment (azasypkin) [inline apps/sms/views/shared/js/smil.js]: nit: looks like we don't need ```activeReaders``` anymore (+ in the place where it's initialized)

File: apps/sms/views/shared/js/smil.js (hunk 1 above)
Comment (azasypkin) [inline apps/sms/views/shared/js/smil.js]: I think we need separate bug to refactor this file including better error handling (I believe not mentored one, as it's not that trivial to make everything right :)).

File: apps/sms/views/shared/test/unit/activity_handler_test.js (hunk 2 above)
Comment (azasypkin) [inline apps/sms/views/shared/test/unit/activity_handler_test.js]: Thanks for the tests you've added!

File: apps/sms/views/conversation/js/conversation.js (hunk 5 above)
Comment (julienw) [inline apps/sms/views/conversation/js/conversation.js]: ah yeah, I meant to remove it !

File: apps/sms/views/conversation/js/conversation.js (hunk 4 above)
Comment (julienw) [inline apps/sms/views/conversation/js/conversation.js]: oki !

File: apps/sms/views/shared/js/smil.js (hunk 7 above)
Comment (julienw) [inline apps/sms/views/shared/js/smil.js]: good point, I started removing them and then stopped :)

File: apps/sms/views/shared/js/smil.js (hunk 1 above)
Comment (julienw) [inline apps/sms/views/shared/js/smil.js]: will file

File: apps/sms/desktop-mock/navigator_moz_mobilemessage.js (hunk 3 above)
Comment (julienw) [inline apps/sms/desktop-mock/navigator_moz_mobilemessage.js]: That's because I tested forwarding a MMS in Firefox to see how this was working with my change, and there was a bug in the desktop mock -- that I fixed then.

So not directly related, but I needed this to test the change in Firefox.

Comment (julienw) [inline apps/sms/desktop-mock/navigator_moz_mobilemessage.js]: because we use the promise returned by getMessage for the forward feature.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=cf77d1469ef5558eb1b9cc7145301c74003a25d0

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8949ab1cd473fd7472fdaab0b4fdfdc5e752fe52

File: apps/sms/views/shared/js/smil.js (hunk 6 above)
Comment (julienw) [inline apps/sms/views/shared/js/smil.js]: filed https://bugzilla.mozilla.org/show_bug.cgi?id=1185555

Comment (azasypkin) [inline apps/sms/views/shared/js/smil.js]: What a beautiful bug number! :)
